API/ev3_timer.c

```diff
@@ -12,268 +16,552 @@
+        {
+            // we found it, move succeeding elements an element back
+            memmove(&pBlock->elements[i],
+                    &pBlock->elements[i + 1],
+                    pBlock->count - i - 1);
+            // reduce count by one
+            pBlock->count -= 1;
+            // and return success
+            return true;
+        }
+    }
+    // no element found, removal failed
+    errno = ENOENT;
+    return false;
 }
 
-#endif
+bool tmArrayMirror(CallbackBlock *pDest, CallbackBlock *pSrc)
+{
+    // let's make sure there's enough room to copy the array
+    if (!tmArrayEnsureSize(pDest, pSrc->count))
+        return false;
+
+    // copy trivial fields (but do NOT copy the capacity field)
+    pDest->enabled = pSrc->enabled;
+    pDest->timer_fd = pSrc->timer_fd;
+    pDest->count = pSrc->count;
+
+    // copy contents of the elements array
+    memcpy(pDest->elements,
+           pSrc->elements,
+           pSrc->count * sizeof(TimerCallback));
+    return true;
+}
 
-unsigned long long TimerGetUS()
+bool tmArrayEnsureReserve(CallbackBlock *pBlock, size_t minReserve)
 {
-  struct timeval tv;
-  gettimeofday(&tv, 0);
-  return (((unsigned long long)tv.tv_sec) * 1000000) + tv.tv_usec;
+    return tmArrayEnsureSize(pBlock, pBlock->count + minReserve);
 }
 
-unsigned long long TimerGetMS()
+bool tmArrayEnsureSize(CallbackBlock *pBlock, size_t minSize)
 {
-  struct timeval tv;
-  gettimeofday(&tv, 0);
-  return (((unsigned long long)tv.tv_sec) * 1000) + (tv.tv_usec / 1000);
+    size_t newCapacity = pBlock->capacity;
+    // initial allocation size
+    if (newCapacity == 0)
+        newCapacity = INITIAL_CALLBACKS;
+
+    // expand until satisfied
+    while (newCapacity < minSize)
+        newCapacity *= 2;
+
+    // however skip if no change is necessary
+    if (newCapacity == pBlock->capacity)
+        return true;
+
+    // let's try to realloc the array; abort if unsuccessful
+    TimerCallback *newArray = realloc(pBlock->elements, newCapacity * sizeof(TimerCallback));
+    if (newArray == NULL)
+        return false;
+
+    // successful; write the results
+    pBlock->elements = newArray;
+    pBlock->capacity = newCapacity;
+    return true;
 }
 
-unsigned long long TimerGetCS()
+// ENABLE & DISABLE TIMERS
+
+bool tmEnablePeriod(TimerInterval interval)
 {
-  return TimerGetMS() / 10;
+    // check if we are not done already
+    if (tcb.callbacksFor[interval].enabled)
+        return true;
+
+    // otherwise fill the timer parameter struct (arms the timer)
+    int millis = tmGetIntervalMilliseconds(interval);
+    int seconds = millis / 1000;
+    int nanos = (millis % 1000) * 1000000;
+
+    struct itimerspec params = {
+            .it_value = {.tv_sec = seconds, .tv_nsec = nanos,},
+            .it_interval = {.tv_sec = seconds, .tv_nsec = nanos,}
+    };
+
+    // push the result to kernel
+    RETURN_ON_FAIL(timerfd_settime(tcb.callbacksFor[interval].timer_fd, 0, &params, NULL),

```

Comment:
If you must use these macros, please use an intermediary variable, so it stays a little readable.

Reply:
Oki, I will change it.

Reply:
Resolved in 620130dbedef6e00dd6d2c55d062508f6af9967a :+1: